lib/spack/spack/binary_distribution.py

```
@@ -629,17 +636,32 @@ def tarball_directory_name(spec):
     Return name of the tarball directory according to the convention
     <os>-<architecture>/<compiler>/<package>-<version>/
     """
-    return spec.format_path("{architecture}/{compiler.name}-{compiler.version}/{name}-{version}")
+    if spec.original_spec_format() < 5:
+        compiler = spec.annotations.compiler_node_attribute
+        assert compiler is not None, "a compiler spec is expected"
```

Is this an assertion or should we throw an error for invalid format?   It doesn't seem like something we can guarantee in the code -- the inputs are essentially arbitrary.

---

This is just an assertion, and the attribute should be there by construction when we read specs from previous formats. Also, as a side note, this shouldn't matter anymore when https://github.com/spack/spack/pull/48713 is merged.

---

> should be there by construction when we read specs from previous formats

My point is really that you can't assert this. Yes, previous versions of Spack *should* have written this, but you don't know that you're not reading corrupt input -- e.g. a corrupt binary.  In that case the issue isn't Spack; it's the file, so the right response is some invalid format exception, not as assertion.

---

I checked, and these modifications are not needed if we merge #48713 first

---

What about now that this is likely to go in first?